File: algorithm/46. Permutations/46.py
class Solution(object):
    def permute(self, nums):
        """
        :type nums: List[int]
        :rtype: List[List[int]]
        """
        def permuteRecu( result, used, cur, num):
            if len(cur) == len(num):
                result.append(cur + [])
                return
            for i in xrange(len(num)):
                if not used[i]:
                    used[i] = True
                    cur.append(num[i])
                    permuteRecu(result, used, cur, num)
                    cur.pop()
                    used[i] = False
        result = []
        used = [False] * len(nums)
        permuteRecu(result, used, [], nums)
        return result

File: algorithm/46. Permutations/46.cpp
class Solution {
public:
    vector<vector<int>> permute(vector<int>& nums) {
        vector<vector<int>> result;
        if(nums.size()==0)
            return result;
        vector<int> visited;
        DFS(nums, visited, result);
        return result;
    }
    void DFS(vector<int> &num,vector<int> &visited,vector<vector<int> > &result)
    {
        if(num.size() == visited.size()){
            result.push_back(visited);
            return;
        }//if
        vector<int>::iterator isVisited;
        for(int i = 0;i < num.size();i++){
            // 判断num[i]是否已经访问过
           isVisited = find(visited.begin(),visited.end(),num[i]);
           // 如果没有访问过
           if(isVisited == visited.end()){
               visited.push_back(num[i]);
               DFS(num,visited,result);
               visited.pop_back();
           }//if
        }//for
    }
};
